### PR TITLE
ports/esp32: add FLASHSIZE_4MB variant to ESP32_GENERIC_S3

### PR DESCRIPTION
**Pull request by thisiseth**
Added a 4MiB flash partitioning variant for ESP32S3: 

don't know how many 4MiB discrete flash board are out there, but there are some with the ESP32-S3FH4R2 chip with embedded 4MiB flash, like [waveshare ESP32-S3 Mini](https://www.waveshare.com/esp32-s3-zero.htm) (also known as ESP32-S3-Zero), which do not work with 8MiB partitions -- the filesystem partition is missing entirely

tested on the mentioned above waveshare ESP32-S3 Mini w/ESP32-S3FH4R2

### CODE REVIEW

**Comment by thisiseth** [comment]
related issue: https://github.com/micropython/micropython/issues/13723

File: ports/esp32/boards/ESP32_GENERIC_S3/board.json [hunk 1]
```diff
@@ -18,6 +18,7 @@
     "url": "https://www.espressif.com/en/products/modules",
     "vendor": "Espressif",
     "variants": {
-        "SPIRAM_OCT": "Support for Octal-SPIRAM"
+        "SPIRAM_OCT": "Support for Octal-SPIRAM",
+        "FLASHSIZE_4MB": "4MiB flash size"
```
Comment (dpgeorge) [inline ports/esp32/boards/ESP32_GENERIC_S3/board.json:22]: Please use `"FLASH_4M": "4MiB flash"`, which matches the existing `ports/esp8266/boards/ESP8266_GENERIC/board.json`.

File: ports/esp32/boards/ESP32_GENERIC_S3/board.md [hunk 2]
```diff
@@ -1,7 +1,9 @@
 The following files are firmware that should work on most ESP32-S3-based
-boards with 8MiB of flash, including WROOM and MINI modules.
+boards with 4/8MiB of flash, including WROOM and MINI modules.
 
 This firmware supports configurations with and without SPIRAM (also known as
 PSRAM) and will auto-detect a connected SPIRAM chip at startup and allocate
 the MicroPython heap accordingly. However if your board has Octal SPIRAM, then
 use the "spiram-oct" variant.
+
+If your board has 4MiB flash (including ESP32-S3FH4R2 based ones with embedded flash), the use the "flashsize-4mb" build.
```
Comment (dpgeorge) [inline ports/esp32/boards/ESP32_GENERIC_S3/board.md:9]: The firmware will have the prefix `ESP32_GENERIC_S3-FLASH_4M-`.

File: ports/esp32/boards/ESP32_GENERIC_S3/mpconfigboard.cmake [hunk 3]
```diff
@@ -19,3 +19,14 @@ if(MICROPY_BOARD_VARIANT STREQUAL "SPIRAM_OCT")
...
 endif()
+
+if(MICROPY_BOARD_VARIANT STREQUAL "FLASHSIZE_4MB")
+    set(SDKCONFIG_DEFAULTS
+        ${SDKCONFIG_DEFAULTS}
+        boards/ESP32_GENERIC_S3/sdkconfig.flash_4mb
+    )
+
+    list(APPEND MICROPY_DEF_BOARD
+        MICROPY_HW_BOARD_NAME="Generic ESP32S3 module with 4MiB flash size"
```
Comment (dpgeorge) [inline ports/esp32/boards/ESP32_GENERIC_S3/mpconfigboard.cmake:30]: I don't think the board name needs to change.  The flash size doesn't really need to be advertised at the REPL banner.

File: ports/esp32/boards/ESP32_GENERIC_S3/board.json (hunk 1 above)
Comment (thisiseth) [reply ports/esp32/boards/ESP32_GENERIC_S3/board.json]: thank you for your review!
changed flashsize_4mb to flash_4m

File: ports/esp32/boards/ESP32_GENERIC_S3/mpconfigboard.cmake (hunk 3 above)
Comment (thisiseth) [reply ports/esp32/boards/ESP32_GENERIC_S3/mpconfigboard.cmake]: agreed, removed custom board name and renamed sdkconfig.flash_4mb to sdkconfig.flash_4m for consistency

**Comment by dpgeorge** [comment]
Thanks for the contribution!

**Pull request merged**